UnitTestFrameworkPkg/GoogleTest.md

```diff
@@ -0,0 +1,1106 @@
+2. `MOCK_FUNCTION_DEFINITION(MOCK, FUNC, NUM_ARGS, CALL_TYPE)`
+3. `MOCK_FUNCTION_INTERNAL_DECLARATION(RET_TYPE, FUNC, ARGS)`
+4. `MOCK_FUNCTION_INTERNAL_DEFINITION(MOCK, FUNC, NUM_ARGS, CALL_TYPE)`
+
+You will notice that there are two sets of macros: one to mock external functions and
+another to mock internal functions. Each set of macros has the exact same arguments, but
+they are used for slightly different use cases. The details of these different use cases
+is described in detail in a later section. For now, we will focus on the usage of the macro
+arguments since that is common between them.
+
+The `MOCK_FUNCTION_DECLARATION` macro is used to declare the mock function in the `.h` file,
+and it takes three arguments: return type, function name, and argument list. The
+`MOCK_FUNCTION_DEFINITION` macro is used to define the mock function in the `.cpp` file,
+and it takes four arguments: name of the interface for the mock, function name, number of
+arguments the function takes, and calling convention type of the function. For example, to
+continue with the `UefiLib` mock example above, the `GetVariable2` and `GetEfiGlobalVariable2`
+functions are declared in `UefiLib.h` as shown below.
+
+```cpp
+EFI_STATUS
+EFIAPI
+GetVariable2 (
+  IN CONST CHAR16    *Name,
+  IN CONST EFI_GUID  *Guid,
+  OUT VOID           **Value,
+  OUT UINTN          *Size OPTIONAL
+  );
+
+EFI_STATUS
+EFIAPI
+GetEfiGlobalVariable2 (
+  IN CONST CHAR16  *Name,
+  OUT VOID         **Value,
+  OUT UINTN        *Size OPTIONAL
+  );
+```
+
+To declare mocks for these functions within the previously created `MockUefiLib` interface,
+the below code would be added to the `MockUefiLib.h` file. Note that the previously added
+interface declaration is also included in the code below for context.
+
+```cpp
+struct MockUefiLib {
+  MOCK_INTERFACE_DECLARATION (MockUefiLib);
+
+  MOCK_FUNCTION_DECLARATION (
+    EFI_STATUS,
+    GetVariable2,
+    (IN CONST CHAR16    *Name,
+     IN CONST EFI_GUID  *Guid,
+     OUT VOID           **Value,
+     OUT UINTN          *Size OPTIONAL)
+    );
+
+  MOCK_FUNCTION_DECLARATION (
+    EFI_STATUS,
+    GetEfiGlobalVariable2,
+    (IN CONST CHAR16  *Name,
+     OUT VOID         **Value,
+     OUT UINTN        *Size OPTIONAL)
+    );
+};
+```
+
+Additionally, the below code would be added into the `MockUefiLib.cpp` file to provide
+the definitions for these mock functions. Again, the previously added interface
+definition is also included in the code below for context.
+
+```cpp
+MOCK_INTERFACE_DEFINITION(MockUefiLib);
+
+MOCK_FUNCTION_DEFINITION(MockUefiLib, GetVariable2, 4, EFIAPI);
+MOCK_FUNCTION_DEFINITION(MockUefiLib, GetEfiGlobalVariable2, 3, EFIAPI);
+```
+
+That concludes the basic overview on how to use the macros, but a more detailed
+description on how to name the mocks, where to put the files, how to build the
+mocks, and how to use the mocks is described in detail later.
+
+### SubhookLib

```

Comment:
If `subhook` still only support x86, I think that should be called out in this section.